nmigen/build/generic_platform.py

```diff
@@ -0,0 +1,161 @@
+from collections import defaultdict, OrderedDict
+
+from .. import *
+from ..hdl.rec import *
+from ..lib.io import *
+
+from .dsl import *
+from .tools import *
+
+
+__all__ = ["ConstraintError", "ConstraintManager"]
+
+
+class ConstraintError(Exception):
+    pass
+
+
+class ConstraintManager:
+    def __init__(self, resources):
+        self.available    = defaultdict(dict)
```

Comment:
First, this should be an OrderedDict too.

Second, a much nicer logic would be to have two dicts, `resources` and `requested`. In the first dict would be contained all resources ever registered with this constraint manager. In the second dict, there would be contained resources requested from this constraint manager.

The reasons are twofold. First, oMigen has a problem that if a resource is requested twice, it acts the second time as if the resource has never existed in the first place. This is very confusing. Second, with the current implementation, you can request, add, request, add, ... the same resource any amount of times.